# <center>  PCA Computation </center>

# Table of Content
# [I. Introduction](#Intro)
# [II. Problem](#Prob) 
# [III. Solution](#Sol)
## [1. Eigenvalues](#Sol1)
## [2. Eigenvectors](#Sol2)
## [3. Principal Compents](#Sol3)
## [4. Check results with Numpy](#Sol4)
# [IV. Conclusions](#Conc)

# I. Introduction<a id='Intro'>

This report concentrates on calculating the eigenvalues, eigenvectors and its principal components for a given covariance matrix by hand.

# II. Problem

Suppose the random variables X1, X2, and X3 have the covariance matrix:

$$ C = 
\begin{pmatrix}
 1 & -2 &  0 \\
-2 &  5 &  0 \\
 0 &  0 &  2 
\end{pmatrix}
$$

Calculate (by hand, show detail work) the eigenvalue-eigenvector pairs and the principal components Y1, Y2 and Y3.

# III. Solution <a id='Sol'>

**Denot:** 

- $\lambda$ is egienvalue.

- **v** is eigenvector.

**Formula**:
- **Cv** = $\lambda$**v** 
- (**C** - $\lambda$**I**)**v** = 0 [1]

The are two cases in this equation - (**C** - $\lambda$**I**) = 0 or **v** = 0. However, **v** = 0 it means eigenvetor = 0, it is meaningless. Therefore, the procedure to solve this problem is that det(**C** - $\lambda$**I**) = 0 is solved to find eigenvalues. Then, apply the eigenvalues to [1] to find eigenvectors.



## 1. Eigenvalues<a id='Sol1'>


det(**C** - $\lambda$**I**) = 0

$$  \mathbf{(C - \lambda I)} =  \begin{pmatrix}
1-\lambda &        -2   & 0 \\
       -2 & 5-\lambda   & 0 \\
        0 &         0   & 2-\lambda \\
\end{pmatrix}
$$


det(**C** - $\lambda$**I**) = $ (1-\lambda)(5-\lambda)(2-\lambda) - (-2)(-2)(2-\lambda) = 0 $

<=> $ (2 - \lambda)(\lambda^2 - 6\lambda  + 1) = 0 $ 

Roots of the equation above are also the value of eigenvalue: they are aranged in order from highest to lowest. 

$\lambda_1 = 3 + 2 \sqrt{2}$

$\lambda_2 = 2 $

$\lambda_3 = 3 - 2 \sqrt{2}$






## 2. Eigenvectors<a id ='Sol2'>

There are 3 eigenvalues, so there are 3 eigenvectors corresponding to each of them. Each $\lambda$ is substituted to the equation [1].

## - $ \mathbf{\lambda_1 = 3 + 2 \sqrt{2}}$

(**C** - $\lambda_1$**I**)**v** = $\begin{bmatrix}
 -2-2\sqrt{2} & -2 & 0 \\
 -2 &  2-2\sqrt{2} & 0  \\
  0 &  0 & -1-2\sqrt{2}
\end{bmatrix} *
\begin{bmatrix}
 v1\\
 v2\\
 v3
\end{bmatrix} = 0
$


- 3rd row $v3 = 0$
- 1st and 2nd rows $v1 =(1 - \sqrt{2}) v2$
- choose $v2 = 1 $ so  $v1 = 1 - \sqrt{2}   $

=> **v** = $\begin{bmatrix}
 1-\sqrt{2}\\
 1\\
 0
\end{bmatrix}$

Similar procedure is applied to remained $\lambda$

## - $ \mathbf{\lambda_1 = 2}$

(**C** - $\lambda_1$**I**)**v** = $\begin{bmatrix}
 -1 & -2 & 0 \\
 -2 &  3 & 0  \\
 0 &  0 &  0
\end{bmatrix} *
\begin{bmatrix}
 v1\\
 v2\\
 v3
\end{bmatrix} = 0
$


- 3rd row $v3 = 1$ (it can be any value and in this case 1 is chosen.)
- 2nd row $v1 =\frac{3}{2} v2$
- 1st row $v2 = 0 $ so $v1 = 0   $

=> **v** = $\begin{bmatrix}
 0\\
 0\\
 1
\end{bmatrix}$


## - $ \mathbf{\lambda_1 = 3 - 2 \sqrt{2}}$

(**C** - $\lambda_1$**I**)**v** = $\begin{bmatrix}
 -2+2\sqrt{2} & -2 & 0 \\
 -2 `&  2+2\sqrt{2} & 0  \\
 0 &  0 & -1+2\sqrt{2}
\end{bmatrix} *
\begin{bmatrix}
 v1\\
 v2\\
 v3
\end{bmatrix} = 0
$


- 3rd row $v3 = 0$
- 1st and 2nd rows $v1 =(1 + \sqrt{2}) v2$
- choose $v2 = 1 $ so $v1 = 1 + \sqrt{2}   $

=> **v** = $\begin{bmatrix}
 1+ \sqrt{2}\\
 1\\
 0
\end{bmatrix}$

**=> Results are:**

$\mathbf{v1}=\begin{bmatrix}
 1-\sqrt{2}\\
 1\\
 0
\end{bmatrix},
\mathbf{v2}=\begin{bmatrix}
 0\\
 0\\
 1
\end{bmatrix},
\mathbf{v3}=\begin{bmatrix}
 1+\sqrt{2}\\
 1\\
 0
\end{bmatrix}
$ with $\lambda_1=3+2\sqrt{2},\lambda_2=2, and \lambda_3=3-2\sqrt{2}$ respectively

<font color ='red'> The eigenvectors have to be unit length to avoid multiple solutions; therefore, they are divided by its Euclidean length  </font>

**=>Final Results are:**

$\mathbf{v1}=\begin{bmatrix}
 -0.3827\\
 0.9239\\
 0
\end{bmatrix},
\mathbf{v2}=\begin{bmatrix}
 0\\
 0\\
 1
\end{bmatrix},
\mathbf{v3}=\begin{bmatrix}
 0.9239\\
 0,3827\\
 0
\end{bmatrix}
$ with $\lambda_1=3+2\sqrt{2},\lambda_2=2, and \lambda_3=3-2\sqrt{2}$ respectively

## 3. Principal Components <a id='Sol3'>

The principal components are:

**PC1:**  
$\mathbf{v1}=\begin{bmatrix}
 -0.3827\\
 0.9239\\
 0
\end{bmatrix}$,
$\lambda_1=3+2\sqrt{2}$

**PC2:**  $\mathbf{v2}=\begin{bmatrix}
 0\\
 0\\
 1
\end{bmatrix},$
$\lambda_2=2$

**PC3:** 
$\mathbf{v3}=\begin{bmatrix}
 0.9239\\
 0.3827\\
 0
\end{bmatrix}$,
$\lambda_3=3-2\sqrt{2}$

## 4. Check result with Numpy <a id='Sol3'>

## Import Libary

In [1]:
import numpy as np
from numpy import linalg as LA

C = np.array([[1, -2, 0],
              [-2, 5, 0],
              [0, 0, 2]])
eigVal, eigVect=LA.eig(C)
print('eigenvalues are:\n', *eigVal)
print('eigenvectors are \n',eigVect)

eigenvalues are:
 0.17157287525381015 5.82842712474619 2.0
eigenvectors are 
 [[-0.92387953  0.38268343  0.        ]
 [-0.38268343 -0.92387953  0.        ]
 [ 0.          0.          1.        ]]




$\mathbf{v1}=\begin{bmatrix}
 -0.3827\\
 0.9239\\
 0
\end{bmatrix},
\mathbf{v2}=\begin{bmatrix}
 0\\
 0\\
 1
\end{bmatrix},
\mathbf{v3}=\begin{bmatrix}
 0.9239\\
 0,3827\\
 0
\end{bmatrix}
$ with $\lambda_1=3+2\sqrt{2}=5.8284,\lambda_2=2, and \lambda_3=3-2\sqrt{2}=0.1716$ respectively.

**=> The reuslts are completely correct because** <font color ='red'> the eigenvectors have to be unit length to avoid multiple solution  </font>
 
 **v1** (by hand) = -1* **v1** (by function) 
 
 **v3** (by hand) = -1* **v3** (by function) 

# IV. Conclusions<a id='Conc'>

$$ C = 
\begin{pmatrix}
 1 & -2 &  0 \\
-2 &  5 &  0 \\
 0 &  0 &  2 
\end{pmatrix}
$$

There are three eigenvalues: $\lambda_1=3+2\sqrt{2}, \lambda_2=2, \lambda_3=3-2\sqrt{2} $

There are three correspoing eigenvectors:

$\mathbf{v1}=\begin{bmatrix}
 -0.3827\\
 0.9239\\
 0
\end{bmatrix},
\mathbf{v2}=\begin{bmatrix}
 0\\
 0\\
 1
\end{bmatrix},
\mathbf{v3}=\begin{bmatrix}
 0.9239\\
 0,3827\\
 0
\end{bmatrix}
$

The principal components are:
 
$\begin{pmatrix}
 -0.3827 & 0 & 0.9239\\
 0.9239  & 0 & 0.3827\\
 0       & 1 & 0
\end{pmatrix}$,
$\lambda_1, \lambda_2, \lambda_3$
